pkg/oci/client/oci.go

```diff
@@ -413,8 +425,79 @@ func (c *client) GetInstanceByNodeName(nodeName string) (*core.Instance, error)
 	return instance, err
 }
 
+func (c *client) waitForVolumeAvailable(volumeID *string, timeout time.Duration) error {
+
+	isVolumeReady := func() (bool, error) {
+		getVolumeRequest := core.GetVolumeRequest{VolumeId: volumeID}
+		ctx, cancel := context.WithTimeout(c.ctx, c.timeout)
+		defer cancel()
+
+		getVolumeResponse, err := c.block.GetVolume(ctx, getVolumeRequest)
+		if err != nil {
+			return false, err
+		}
+
+		if getVolumeResponse.LifecycleState == core.VolumeLifecycleStateAvailable {
+			return true, nil
+		}
+
+		if getVolumeResponse.LifecycleState == core.VolumeLifecycleStateFaulty {
+			return false, fmt.Errorf("Failed to provision volume. Volume: %v is in Faulty state", *volumeID)
+		}
+
+		if getVolumeResponse.LifecycleState == core.VolumeLifecycleStateTerminated {
+			return false, fmt.Errorf("Failed to provision volume. Volume: %v is in Terminated state", *volumeID)
+		}
+
+		if getVolumeResponse.LifecycleState == core.VolumeLifecycleStateTerminating {
+			return false, fmt.Errorf("Failed to provision volume. Volume: %v is in Terminating state", *volumeID)
+		}
+
+		return false, nil
+	}
+
+	ready, err := isVolumeReady()
+	if err != nil {
+		return err
+	}
+	if ready {
+		return nil
+	}
+
+	//Not ready so we'll kick off a polling loop to wait for readyness or failure...
+	tickduration := time.Second * 5
+	timeoutticker := time.NewTicker(timeout)
+	defer timeoutticker.Stop()
+
+	ticker := time.NewTicker(tickduration)
+	defer ticker.Stop()
+
+	// Keep trying until we're timed out or got a result or got an error
+	for {
+		select {
+		case <-timeoutticker.C:
+			return fmt.Errorf("Timed out waiting for volume %v to become %s", *volumeID, core.VolumeLifecycleStateAvailable)
+
+		case <-ticker.C:
+			ready, err := isVolumeReady()
+			if err != nil {
+				return err
+			}
+			if ready {
+				return nil
+			}
+		}
+	}
+}
+
 // AttachVolume attaches a block storage volume to the specified instance.
 func (c *client) AttachVolume(instanceId, volumeId string) (core.VolumeAttachment, int, error) {
+	// Wait for volume to be available
+	err := c.waitForVolumeAvailable(&volumeId, time.Minute*3)
```

Comment:
3 minutes might be too long here for the timeout. We could likely drop it to a minute or so.